lib/appoptics_apm/oboe_init_options.rb

```diff
@@ -0,0 +1,135 @@
+
+      # the reporter to be used (ssl, upd, file, null)
+      # collector endpoint (reporter=ssl), udp address (reporter=udp), or file path (reporter=file)
+      reporter_and_host # sets @reporter and @host
+
+      # the service key
+      @service_key = ENV['APPOPTICS_SERVICE_KEY'] || AppOpticsAPM::Config[:service_key] || ''
+      # path to the SSL certificate (only for ssl)
+      @trusted_path = ENV['APPOPTICS_TRUSTEDPATH'] || ''
+      # size of the message buffer
+      @buffer_size = (ENV['APPOPTICS_BUFSIZE'] || -1).to_i
+      # flag indicating if trace metrics reporting should be enabled (default) or disabled
+      @trace_metrics = (ENV['APPOPTICS_TRACE_METRICS'] || -1).to_i
+      # the histogram precision (only for ssl)
+      @histogram_precision = (ENV['APPOPTICS_HISTOGRAM_PRECISION'] || -1).to_i
+      # custom token bucket capacity
+      @token_bucket_capacity = (ENV['APPOPTICS_TOKEN_BUCKET_CAPACITY'] || -1).to_i
+      # custom token bucket rate
+      @token_bucket_rate = (ENV['APPOPTICS_TOKEN_BUCKET_RATE'] || -1).to_i
+      # use single files in file reporter for each event
+      @file_single = (ENV['APPOPTICS_REPORTER_FILE_SINGLE'].to_s.downcase == 'true') ? 1 : 0
+    end
+
+    def re_init # for testing with changed ENV vars
+      initialize
+    end
+
+    def array_for_oboe
+      [
+        @hostname_alias,
+        @debug_level,
+        @log_file_path,
+        @max_transactions,
+        @max_flush_wait_time,
+        @events_flush_interval,
+        @event_flush_batch_size,
+
+        @reporter,
+        @host,
+        @service_key,
+        @trusted_path,
+        @buffer_size,
+        @trace_metrics,
+        @histogram_precision,
+        @token_bucket_capacity,
+        @token_bucket_rate,
+        @file_single
+      ]
+    end
+
+    def service_key_ok?
+      # only required for ssl reporter
+      if @reporter == 'ssl' && @service_key !~ /^[0-9a-fA-F]{64}:[-.:_?\\\/\w ]{1,255}$/
+        AppOpticsAPM.logger.warn '[appoptics_apm/oboe_options] APPOPTICS_SERVICE_KEY problem. No service name or api token in wrong format. Cannot submit data.'
+        return false
+      end
+      true
+    end
+
+    private
+
+    def reporter_and_host
+
+      @reporter = ENV['APPOPTICS_REPORTER'] || 'ssl'
+      # override with 'file', e.g. when running tests
+      @reporter = 'file' if ENV.key?('APPOPTICS_GEM_TEST')
+
+      @host = ''
+      case @reporter
+      when 'ssl', 'file'
+        @host = ENV['APPOPTICS_COLLECTOR'] || ''
+      when 'udp'
+        @host = ENV['APPOPTICS_COLLECTOR'] ||
+                "#{AppOpticsAPM::Config[:reporter_host]}:#{AppOpticsAPM::Config[:reporter_port]}" || ''
+        # TODO decide what to do
+        # ____ AppOpticsAPM::Config[:reporter_host] and
+        # ____ AppOpticsAPM::Config[:reporter_port] were moved here from
+        # ____ oboe_metal.rb and are not documented anywhere
+      when 'null'
+        @host = ''
+      end
+
```

Comment:
no need for `"#{AppOpticsAPM::Config[:reporter_host]}:#{AppOpticsAPM::Config[:reporter_port]}" || ''` the trailing `|| ''` because the substituted string cannot possibly be false (as far as my knowledge of ruby goes).

Reply:
right!